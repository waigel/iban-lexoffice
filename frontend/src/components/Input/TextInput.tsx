import React, {useEffect, useState} from 'react';
import {classNames} from '../../utils/classNames';

export interface TextInputProps
  extends React.HtmlHTMLAttributes<HTMLInputElement> {
  handleValueChange?: (value: string) => void;
  handleEnterPress?: () => void;
  defaultValue?: string;
}

export const TextInput = (props: TextInputProps) => {
  const {handleValueChange, defaultValue, handleEnterPress} = props;
  const [value, setValue] = useState<string>(defaultValue ?? '');

  useEffect(() => {
    defaultValue && setValue(defaultValue);
  }, [defaultValue]);

  const domProps = {...props};
  delete domProps.defaultValue;
  delete domProps.handleEnterPress;
  delete domProps.handleValueChange;

  return (
    <input
      {...domProps}
      className={classNames(
        'bg-transparent border rounded-xl px-4 py-2 text-center tracking-widest',
        props.className
      )}
      onKeyDown={event => {
        if (event.key === 'Enter') {
          handleEnterPress && handleEnterPress();
        }
      }}
      onChange={event => {
        const changedValue = event.target.value;
        setValue(changedValue);
        handleValueChange && handleValueChange(changedValue);
      }}
      value={value}
    />
  );
};
