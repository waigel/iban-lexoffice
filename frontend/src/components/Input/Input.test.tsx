import React from 'react';
import TestRenderer, {act} from 'react-test-renderer';
import {TextInput} from './TextInput';

it('should test the TextInput component is rendered correctly', () => {
  const tree = TestRenderer.create(<TextInput />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('should test the TextInput onChange event is called correctly', () => {
  const onChange = jest.fn();
  const tree = TestRenderer.create(<TextInput handleValueChange={onChange} />);
  act(() => {
    const input = tree.root.findByType('input');
    input.props.onChange({target: {value: 'test'}});
  });
  expect(onChange).toBeCalledWith('test');
});

it('should test the TextInput handleEnterPress event is called correctly', () => {
  const handleEnterPress = jest.fn();
  const tree = TestRenderer.create(
    <TextInput handleEnterPress={handleEnterPress} />
  );
  act(() => {
    const input = tree.root.findByType('input');
    input.props.onKeyDown({key: 'Enter'});
  });
  expect(handleEnterPress).toBeCalled();
});

it('should test the TextInput handleEnterPress event is not called when key is not Enter', () => {
  const handleEnterPress = jest.fn();
  const tree = TestRenderer.create(
    <TextInput handleEnterPress={handleEnterPress} />
  );
  act(() => {
    const input = tree.root.findByType('input');
    input.props.onKeyDown({key: 'ALT'});
  });
  expect(handleEnterPress).not.toBeCalled();
});

it('should test the TextInput defaultValue is set correctly', () => {
  const tree = TestRenderer.create(<TextInput defaultValue={'test'} />);
  const input = tree.root.findByType('input');
  expect(input.props.value).toEqual('test');
});
