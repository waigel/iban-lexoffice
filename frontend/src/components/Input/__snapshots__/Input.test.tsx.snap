// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should test the TextInput component is rendered correctly 1`] = `
<input
  className="bg-transparent border rounded-xl px-4 py-2 text-center tracking-widest"
  onChange={[Function]}
  onKeyDown={[Function]}
  value=""
/>
`;
